File: golang/bulk_command_execution/v0_bulk_command_api/v0_bulk_commands.go
package v0_bulk_command_api

import (
	"encoding/json"
	"github.com/kurtosis-tech/kurtosis-client/golang/core_api_bindings"
	"github.com/palantir/stacktrace"
)

// ====================================================================================================
//                                   Command Arg Deserialization Visitor
// ====================================================================================================

// Visitor that will be used to deserialize command args into
type cmdArgDeserializingVisitor struct {
	bytesToDeserialize []byte
	deserializedCommandArgs interface{}
}

func newCmdArgDeserializingVisitor(bytesToDeserialize []byte) *cmdArgDeserializingVisitor {
	return &cmdArgDeserializingVisitor{bytesToDeserialize: bytesToDeserialize}
}

func (visitor cmdArgDeserializingVisitor) VisitRegisterService() error {
	args := &core_api_bindings.RegisterServiceArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the register service args")
	}
	visitor.deserializedCommandArgs = args
	return nil
}

func (visitor cmdArgDeserializingVisitor) VisitGenerateFiles() error {
	args := &core_api_bindings.GenerateFilesArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the generate files args")
	}
	visitor.deserializedCommandArgs = args
	return nil
}

func (visitor cmdArgDeserializingVisitor) VisitStartService() error {
	args := &core_api_bindings.StartServiceArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the start service args")
	}
	visitor.deserializedCommandArgs = args
	return nil
}

func (visitor cmdArgDeserializingVisitor) VisitRemoveService() error {
	args := &core_api_bindings.RemoveServiceArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the remove service args")
	}
	visitor.deserializedCommandArgs = args
	return nil
}

func (visitor cmdArgDeserializingVisitor) VisitRepartition() error {
	args := &core_api_bindings.RepartitionArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the repartition service args")
	}
	visitor.deserializedCommandArgs = args
	return nil
}

func (visitor cmdArgDeserializingVisitor) VisitExecCommand() error {
	args := &core_api_bindings.ExecCommandArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the exec command args")
	}
	visitor.deserializedCommandArgs = args
	return nil
}

func (visitor cmdArgDeserializingVisitor) VisitWaitForEndpointAvailability() error {
	args := &core_api_bindings.WaitForEndpointAvailabilityArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the endpoint availability-waiting args")
	}
	visitor.deserializedCommandArgs = args
	return nil
}

func (visitor cmdArgDeserializingVisitor) VisitExecuteBulkCommands() error {
	args := &core_api_bindings.ExecuteBulkCommandsArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the bulk command execution args")
	}
	visitor.deserializedCommandArgs = args
	return nil
}

func (visitor cmdArgDeserializingVisitor) GetDeserializedCommandArgs() interface{} {
	return visitor.deserializedCommandArgs
}

// ====================================================================================================
//                                        Serializable Command
// ====================================================================================================

// Used for serializing
type V0SerializableCommand struct {
	Type V0CommandType `json:"type"`

	// The only allowed objects here are from the bindings generated from the .proto file
	Args interface{}	`json:"args"`
}

// A V0SerializableCommand knows how to deserialize itself, thanks to the "type" tag
func (cmd *V0SerializableCommand) UnmarshalJSON(bytes []byte) error {
	interstitialStruct := struct {
		Type      V0CommandType   `json:"type"`
		ArgsBytes json.RawMessage `json:"args"`
	}{}
	if err := json.Unmarshal(bytes, &interstitialStruct); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the bytes into a command")
	}

	cmdArgDeserializingVisitor := newCmdArgDeserializingVisitor(interstitialStruct.ArgsBytes)
	if err := cmd.Type.AcceptVisitor(cmdArgDeserializingVisitor); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing command with the following JSON: %v", string(bytes))
	}
	cmd.Args = cmdArgDeserializingVisitor.GetDeserializedCommandArgs()

	return nil
}


// ====================================================================================================
//                                   Bulk Commands Package
// ====================================================================================================

type V0BulkCommands struct {
	Commands []V0SerializableCommand `json:"commands"`
}

File: typescript/lib/bulk_command_execution/v0_bulk_command_api/v0_bulk_commands.ts
import { LoadLambdaArgs, ExecCommandArgs, ExecuteBulkCommandsArgs, ExecuteLambdaArgs, GenerateFilesArgs, LoadStaticFilesArgs, RegisterServiceArgs, RemoveServiceArgs, RepartitionArgs, StartServiceArgs, WaitForEndpointAvailabilityArgs } from "../../../kurtosis_core_rpc_api_bindings/api_container_service_pb";
import { V0CommandTypeVisitor, V0CommandType } from "./v0_command_types";
import { ok, err, Result } from "neverthrow";
import * as protobuf from "google-protobuf";

// ====================================================================================================
//                                   Command Arg Deserialization Visitor
// ====================================================================================================

// Visitor that will be used to deserialize command args into
class CmdArgDeserializingVisitor implements V0CommandTypeVisitor {
    private readonly bytesToDeserialize: string;
    private deserializedCommandArgsPtr: protobuf.Message;

    constructor (bytesToDeserialize: string) {
        this.bytesToDeserialize = bytesToDeserialize;
    }

    public visitLoadLambda(): Result<null, Error> {
        let args: LoadLambdaArgs;
       
        try {
            args = JSON.parse(this.bytesToDeserialize);
        } catch(jsonErr) {
            return err(jsonErr);
        }

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitExecuteLambda(): Result<null, Error> {
    	let args: ExecuteLambdaArgs;
        
        try {
            args = JSON.parse(this.bytesToDeserialize);
        } catch(jsonErr) {
            return err(jsonErr);
        }

    	this.deserializedCommandArgsPtr = args;
    	return ok(null);
    }

    public visitRegisterService(): Result<null, Error> {
        let args: RegisterServiceArgs;
        
        try {
            args = JSON.parse(this.bytesToDeserialize);
        } catch(jsonErr) {
            return err(jsonErr);
        }

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitGenerateFiles(): Result<null, Error> {
        let args: GenerateFilesArgs;
        
        try {
            args = JSON.parse(this.bytesToDeserialize);
        } catch(jsonErr) {
            return err(jsonErr);
        }

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitLoadStaticFiles(): Result<null, Error> {
    	let args: LoadStaticFilesArgs;
    	 
        try {
            args = JSON.parse(this.bytesToDeserialize);
        } catch(jsonErr) {
            return err(jsonErr);
        }

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitStartService(): Result<null, Error> {
        let args: StartServiceArgs;

        try {
            args = JSON.parse(this.bytesToDeserialize);
        } catch(jsonErr) {
            return err(jsonErr);
        }

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitRemoveService(): Result<null, Error> {
    	let args: RemoveServiceArgs;
    	
        try {
            args = JSON.parse(this.bytesToDeserialize);
        } catch(jsonErr) {
            return err(jsonErr);
        }

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitRepartition(): Result<null, Error> {
    	let args: RepartitionArgs;
    	
        try {
            args = JSON.parse(this.bytesToDeserialize);
        } catch(jsonErr) {
            return err(jsonErr);
        }

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitExecCommand(): Result<null, Error> {
        let args: ExecCommandArgs;
     
        try {
            args = JSON.parse(this.bytesToDeserialize);
        } catch(jsonErr) {
            return err(jsonErr);
        }

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitWaitForEndpointAvailability(): Result<null, Error> {
    	let args: WaitForEndpointAvailabilityArgs;
    	
        try {
            args = JSON.parse(this.bytesToDeserialize);
        } catch(jsonErr) {
            return err(jsonErr);
        }

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitExecuteBulkCommands(): Result<null, Error> {
        let args: ExecuteBulkCommandsArgs;

        try {
            args = JSON.parse(this.bytesToDeserialize);
        } catch(jsonErr) {
            return err(jsonErr);
        }

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public getDeserializedCommandArgs(): protobuf.Message {
        return this.deserializedCommandArgsPtr;
    }
}

// ====================================================================================================
//                                        Serializable Command
// ====================================================================================================

// Used for serializing
export class V0SerializableCommand {
    private readonly type: V0CommandType;
    
    // The only allowed objects here are from the bindings generated from the .proto file
    private readonly argsPtr: protobuf.Message;

    constructor (type: V0CommandType, argsPtr: protobuf.Message) {
        this.type = type;
        this.argsPtr = argsPtr;
    } 
       
    public getType(): V0CommandType {
        return this.type;
    }

    public getArgsPtr(): protobuf.Message {
        return this.argsPtr;
    }

}

// // ====================================================================================================
// //                                   Bulk Commands Package
// // ====================================================================================================

export class V0BulkCommands {
    private readonly commands: V0SerializableCommand[];

    constructor (commands: V0SerializableCommand[]) {
        this.commands = commands;
    }
}